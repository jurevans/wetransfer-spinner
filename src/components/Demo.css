.demo {
  margin: 0 auto;
  text-align: center;
  width: 600px;
}

.demo-spinners {
  display: flex;
  flex-wrap: wrap;
}

.demo-buttons {
  margin: 20px 0;
}
