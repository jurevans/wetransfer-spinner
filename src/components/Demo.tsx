import {
  FC,
  ReactElement,
  useCallback,
  useEffect,
  useState,
} from 'react';
import Spinner from './Spinner';
import SVGCounter from './SVGCounter';
import withHTMLCounter from './withHTMLCounter';
import './Demo.css';

const SpinnerWithHTMLCounter = withHTMLCounter(Spinner);

const Demo: FC = (): ReactElement => {
  const [progress, setProgress] = useState(0);
  const [isSpinning, setIsSpinning] = useState(false);
  const [timeoutId, setTimeoutId] = useState(0);

  const toggleIsSpinning = () => {
    setIsSpinning(!isSpinning);
  };

  const spin = useCallback(() => {
    if (progress === 100) {
      setProgress(0);
    } else {
      // Increment progress by a number between 1 and 5, with the result not
      // exceeding 100:
      setProgress(Math.min(progress + Math.floor(Math.random() * 5) + 1, 100));
    }
  }, [progress]);

  useEffect(() => {
    if (isSpinning) {
      const duration = (() => {
        switch(progress) {
          case 0:
            return 1500;
          case 100:
            return 1000;
          default:
            // Multiply 80ms by a random number between 1 and 5
            return (Math.floor(Math.random() * 5) + 1) * 80;
        }
      })();
      const id: number = window.setTimeout(spin, duration);
      setTimeoutId(id);
    }
  }, [progress]);

  const handleStartSpinning = () => {
    toggleIsSpinning();
    spin();
  };

  const handleStopSpinning = () => {
    clearTimeout(timeoutId)
    toggleIsSpinning();
  };

  // Define radius for each demo spinner:
  const radius1 = 175;
  const radius2 = 125;
  const radius3 = 110;
  const radius4 = 300;

  return (
    <div className='demo'>
      <div className='demo-spinners'>
        <Spinner
          progress={progress}
          transitionDuration={0.75}
          rotate={isSpinning}
          radius={radius1}
          size='33%'
        >
          <SVGCounter
            x={radius1}
            y={radius1}
            percentage={progress}
          />
        </Spinner>
        <Spinner
          progress={progress}
          transitionDuration={0.50}
          rotate={isSpinning}
          rotateDuration={1.5}
          radius={radius2}
          size='33%'
        >
          <SVGCounter
            x={radius2}
            y={radius2}
            percentage={progress}
          />
        </Spinner>
        <Spinner
          progress={progress}
          transitionDuration={0.50}
          rotate={isSpinning}
          rotateDuration={1.25}
          radius={radius3}
          size='33%'
          strokeWidth={40}
          progressStyles={{ stroke: '#884499', strokeLinecap: 'unset' }}
        >
          <SVGCounter
            x={radius3}
            y={radius3}
            percentage={progress}
          />
        </Spinner>
        <Spinner
          progress={progress}
          transitionDuration={1.25}
          rotate={isSpinning}
          rotateDuration={4}
          radius={radius4}
          size={radius4}
        >
          <SVGCounter
            x='47%'
            y='47%'
            percentage={progress}
          />
        </Spinner>
        <SpinnerWithHTMLCounter
          progress={progress}
          rotate={isSpinning}
        />
      </div>
      <div className='demo-buttons'>
        {isSpinning
          ? <button onClick={handleStopSpinning}>End</button>
          : <button onClick={handleStartSpinning}>Start</button>}
      </div>
    </div>
  );
};

export default Demo;
