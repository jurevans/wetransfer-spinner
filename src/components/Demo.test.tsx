import React from 'react';
import { fireEvent, render, screen } from '@testing-library/react';
import renderer from 'react-test-renderer';
import Demo from './Demo';

describe('Demo', () => {
  it('Renders correctly', () => {
    const tree = renderer.create(<Demo />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('Renders start button', () => {
    render(<Demo />);
    const startElement = screen.getByText(/start/i);
    expect(startElement).toBeInTheDocument();
  });

  it('Renders the end button once start has been clicked', () => {
    render(<Demo />);
    fireEvent(
      screen.getByText('Start'),
      new MouseEvent('click', {
        bubbles: true,
        cancelable: true,
      }),
    );
    const endButton = screen.getByText(/end/i);
    expect(endButton).toBeInTheDocument();
  });

  it('Renders a non-zero progress when Start and End have been clicked', () => {
    const { container } = render(<Demo />);
    fireEvent(
      screen.getByText('Start'),
      new MouseEvent('click', {
        bubbles: true,
        cancelable: true,
      }),
    );
    fireEvent(
      screen.getByText('End'),
      new MouseEvent('click', {
        bubbles: true,
        cancelable: true,
      }),
    );
    expect(container.getElementsByClassName('percentage-value')[0].innerHTML).not.toBe('0');
  });
});
