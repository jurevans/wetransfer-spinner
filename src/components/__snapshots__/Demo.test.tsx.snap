// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Demo Renders correctly 1`] = `
<div
  className="demo"
>
  <div
    className="demo-spinners"
  >
    <svg
      className="progress-spinner"
      height="33%"
      viewBox="-25 -25 400 400"
      width="33%"
    >
      <circle
        className="progress-circle"
        cx={175}
        cy={175}
        r={175}
        strokeWidth={25}
      />
      <circle
        className="progress-bar"
        cx={175}
        cy={175}
        r={175}
        strokeDasharray={1100}
        strokeWidth={25}
        style={
          Object {
            "animationDuration": "2s",
            "animationPlayState": "paused",
            "strokeDashoffset": 1100,
            "transformOrigin": "175px 175px",
            "transition": "stroke-dashoffset 0.75s ease-out",
          }
        }
      />
      <text
        className="svg-counter"
        dominantBaseline="central"
        textAnchor="middle"
        x={175}
        y={175}
      >
        <tspan
          className="svg-counter-value"
        >
          0
        </tspan>
        <tspan
          className="svg-counter-sup"
          dx={6}
          dy={-14}
        >
          %
        </tspan>
      </text>
    </svg>
    <svg
      className="progress-spinner"
      height="33%"
      viewBox="-25 -25 300 300"
      width="33%"
    >
      <circle
        className="progress-circle"
        cx={125}
        cy={125}
        r={125}
        strokeWidth={25}
      />
      <circle
        className="progress-bar"
        cx={125}
        cy={125}
        r={125}
        strokeDasharray={785}
        strokeWidth={25}
        style={
          Object {
            "animationDuration": "1.5s",
            "animationPlayState": "paused",
            "strokeDashoffset": 785,
            "transformOrigin": "125px 125px",
            "transition": "stroke-dashoffset 0.5s ease-out",
          }
        }
      />
      <text
        className="svg-counter"
        dominantBaseline="central"
        textAnchor="middle"
        x={125}
        y={125}
      >
        <tspan
          className="svg-counter-value"
        >
          0
        </tspan>
        <tspan
          className="svg-counter-sup"
          dx={6}
          dy={-14}
        >
          %
        </tspan>
      </text>
    </svg>
    <svg
      className="progress-spinner"
      height="33%"
      viewBox="-40 -40 310 310"
      width="33%"
    >
      <circle
        className="progress-circle"
        cx={115}
        cy={115}
        r={115}
        strokeWidth={40}
      />
      <circle
        className="progress-bar"
        cx={115}
        cy={115}
        r={115}
        strokeDasharray={723}
        strokeWidth={40}
        style={
          Object {
            "animationDuration": "1.25s",
            "animationPlayState": "paused",
            "stroke": "#884499",
            "strokeDashoffset": 723,
            "strokeLinecap": "unset",
            "transformOrigin": "115px 115px",
            "transition": "stroke-dashoffset 0.5s ease-out",
          }
        }
      />
      <text
        className="svg-counter"
        dominantBaseline="central"
        textAnchor="middle"
        x={115}
        y={115}
      >
        <tspan
          className="svg-counter-value"
        >
          0
        </tspan>
        <tspan
          className="svg-counter-sup"
          dx={6}
          dy={-14}
        >
          %
        </tspan>
      </text>
    </svg>
    <svg
      className="progress-spinner"
      height={300}
      viewBox="-25 -25 650 650"
      width={300}
    >
      <circle
        className="progress-circle"
        cx={300}
        cy={300}
        r={300}
        strokeWidth={25}
      />
      <circle
        className="progress-bar"
        cx={300}
        cy={300}
        r={300}
        strokeDasharray={1885}
        strokeWidth={25}
        style={
          Object {
            "animationDuration": "4s",
            "animationPlayState": "paused",
            "strokeDashoffset": 1885,
            "transformOrigin": "300px 300px",
            "transition": "stroke-dashoffset 1.25s ease-out",
          }
        }
      />
      <text
        className="svg-counter"
        dominantBaseline="central"
        textAnchor="middle"
        x="47%"
        y="47%"
      >
        <tspan
          className="svg-counter-value"
        >
          0
        </tspan>
        <tspan
          className="svg-counter-sup"
          dx={6}
          dy={-14}
        >
          %
        </tspan>
      </text>
    </svg>
  </div>
  <div
    className="demo-buttons"
  >
    <button
      onClick={[Function]}
    >
      Start
    </button>
  </div>
</div>
`;
